File: test-fase1.py
#!/usr/bin/env python3
"""
Zami Girls — Prueba Fase 1: Generación de Rostros
Corre: python3 test-fase1.py
"""

import json, os, random, time, urllib.request, urllib.error, uuid
from pathlib import Path

# ─── CONFIG ───────────────────────────────────────────────────────────────────
COMFYUI_URL = "http://127.0.0.1:8188"   # local dentro de RunPod
OUTPUT_DIR  = Path(__file__).parent / "test-output"
WORKFLOW    = json.loads((Path(__file__).parent / "src/workflows/face-generation.json").read_text())

PARAMS = {
    "ethnicity":   "Hispanic",
    "skin_tone":   "medium",
    "eye_color":   "brown",
    "hair_color":  "dark brown",
    "hair_length": "shoulder-length",
    "expression":  "soft smile",
    "photo_type":  "Studio white background",
    "brief_text":  "beautiful latina woman, elegant, natural",
}

# ─── HELPERS ──────────────────────────────────────────────────────────────────
def rand_seed():
    return random.randint(1, 999_999_999_999_999)

def build_workflow(params, seed):
    wf = json.loads(json.dumps(WORKFLOW))   # deep copy
    wf["78"]["inputs"]["seed"]           = seed
    wf["14"]["inputs"]["noise_seed"]     = rand_seed()
    wf["73"]["inputs"]["seed"]           = rand_seed()
    for k in ["brief_text","photo_type","ethnicity","skin_tone","eye_color",
              "hair_color","hair_length","expression"]:
        if params.get(k):
            wf["82"]["inputs"][k] = params[k]
    return wf

def post_json(url, data):
    body = json.dumps(data).encode()
    req  = urllib.request.Request(url, data=body,
                                  headers={"Content-Type": "application/json"})
    with urllib.request.urlopen(req, timeout=15) as r:
        return json.loads(r.read())

def get_json(url):
    with urllib.request.urlopen(url, timeout=15) as r:
        return json.loads(r.read())

def queue_prompt(workflow):
    data = post_json(f"{COMFYUI_URL}/prompt",
                     {"prompt": workflow, "client_id": str(uuid.uuid4())})
    pid  = data["prompt_id"]
    print(f"  → En cola: {pid}")
    return pid

def poll_result(prompt_id, idx):
    print(f"  [{idx+1}] Generando", end="", flush=True)
    for _ in range(150):
        time.sleep(2)
        try:
            data = get_json(f"{COMFYUI_URL}/history/{prompt_id}")
            if data.get(prompt_id, {}).get("outputs"):
                print(" ✓")
                return data[prompt_id]
        except Exception:
            pass
        print(".", end="", flush=True)
    raise TimeoutError("ComfyUI tardó más de 5 min")

def extract_images(entry):
    imgs = []
    for out in entry.get("outputs", {}).values():
        imgs.extend(out.get("images", []))
    return imgs

def download_image(filename, img_type, subfolder, dest):
    url = (f"{COMFYUI_URL}/view"
           f"?filename={urllib.parse.quote(filename)}"
           f"&type={img_type}"
           f"&subfolder={urllib.parse.quote(subfolder or '')}")
    with urllib.request.urlopen(url, timeout=60) as r:
        dest.write_bytes(r.read())

# ─── MAIN ─────────────────────────────────────────────────────────────────────
import urllib.parse

def main():
    print("\n╔══════════════════════════════════════════╗")
    print("║  ZAMI GIRLS — Prueba Fase 1: Rostros     ║")
    print("╚══════════════════════════════════════════╝\n")

    # Verificar ComfyUI
    print("▸ Verificando ComfyUI...")
    try:
        stats = get_json(f"{COMFYUI_URL}/system_stats")
        vram  = stats.get("system", {}).get("vram_total", 0)
        print(f"  ✓ ComfyUI activo — VRAM total: {vram // 1_000_000_000} GB\n")
    except Exception as e:
        print(f"  ✗ No se puede conectar: {e}")
        print(f"  URL usada: {COMFYUI_URL}")
        print("  Asegúrate de que ComfyUI está corriendo en el pod.\n")
        return

    OUTPUT_DIR.mkdir(exist_ok=True)

    # Encolar 4 generaciones
    print("▸ Encolando 4 generaciones...")
    jobs = []
    for i in range(4):
        seed     = rand_seed()
        workflow = build_workflow(PARAMS, seed)
        pid      = queue_prompt(workflow)
        jobs.append({"pid": pid, "seed": seed, "idx": i})
        time.sleep(0.4)

    # Esperar resultados
    print("\n▸ Esperando resultados (~2 min por imagen)...")
    results = []
    for job in jobs:
        try:
            entry  = poll_result(job["pid"], job["idx"])
            images = extract_images(entry)
            if images:
                img  = images[-1]   # última = upscaled final
                dest = OUTPUT_DIR / f"rostro-{job['idx']+1}-seed{job['seed']}.png"
                download_image(img["filename"], img["type"], img.get("subfolder",""), dest)
                print(f"  ✓ Guardada: {dest.name}")
                results.append({**job, "file": dest, "img": img})
        except Exception as e:
            print(f"\n  ✗ Job {job['idx']+1} falló: {e}")

    # Resumen
    print("\n╔══════════════════════════════════════════╗")
    print("║  RESULTADOS                              ║")
    print("╚══════════════════════════════════════════╝")
    for r in results:
        url = (f"{COMFYUI_URL}/view?filename={urllib.parse.quote(r['img']['filename'])}"
               f"&type={r['img']['type']}")
        print(f"\n  Rostro {r['idx']+1} | seed: {r['seed']}")
        print(f"  Archivo : {r['file']}")
        print(f"  Ver en  : {url}")
    print(f"\n✓ {len(results)}/4 rostros generados en: {OUTPUT_DIR}\n")

if __name__ == "__main__":
    main()

File: test-fase1.js
#!/usr/bin/env node
/**
 * Zami Girls — Prueba Fase 1: Generación de Rostros
 * Corre: node test-fase1.js
 * Requiere Node.js 18+
 */

import fs from 'fs'
import path from 'path'
import { fileURLToPath } from 'url'

const __dir = path.dirname(fileURLToPath(import.meta.url))

// ─── CONFIG ───────────────────────────────────────────────────────────────────
const COMFYUI_URL = 'https://yox2fumbwtg8fb-8188.proxy.runpod.net'
const OUTPUT_DIR  = path.join(__dir, 'test-output')
const WORKFLOW    = JSON.parse(fs.readFileSync(path.join(__dir, 'src/workflows/face-generation.json'), 'utf8'))

// Parámetros del rostro a generar
const PARAMS = {
  ethnicity:   'Hispanic',
  skin_tone:   'medium',
  eye_color:   'brown',
  hair_color:  'dark brown',
  hair_length: 'shoulder-length',
  expression:  'soft smile',
  photo_type:  'Studio white background',
  brief_text:  'beautiful latina woman, elegant, natural',
}

// ─── HELPERS ──────────────────────────────────────────────────────────────────
function randomSeed() {
  return Math.floor(Math.random() * 999999999999999)
}

function buildWorkflow(params, seed) {
  const wf = JSON.parse(JSON.stringify(WORKFLOW))

  // Seeds
  wf['78'].inputs.seed           = seed
  wf['14'].inputs.noise_seed     = randomSeed()
  wf['73'].inputs.seed           = randomSeed()

  // Parámetros faciales en nodo AION (82)
  const allowed = ['brief_text','photo_type','ethnicity','skin_tone','eye_color',
    'hair_color','hair_length','expression','expression_variant']
  for (const k of allowed) {
    if (params[k]) wf['82'].inputs[k] = params[k]
  }

  return wf
}

async function queuePrompt(workflow) {
  const res = await fetch(`${COMFYUI_URL}/prompt`, {
    method:  'POST',
    headers: { 'Content-Type': 'application/json' },
    body:    JSON.stringify({ prompt: workflow, client_id: crypto.randomUUID() }),
  })
  if (!res.ok) throw new Error(`Queue error: ${res.status} ${await res.text()}`)
  const data = await res.json()
  console.log(`  → Prompt en cola: ${data.prompt_id}`)
  return data.prompt_id
}

async function pollResult(promptId, idx) {
  process.stdout.write(`  [${idx+1}] Generando`)
  for (let i = 0; i < 150; i++) {
    await new Promise(r => setTimeout(r, 2000))
    const res  = await fetch(`${COMFYUI_URL}/history/${promptId}`)
    const data = await res.json()
    if (data[promptId]?.outputs) {
      process.stdout.write(' ✓\n')
      return data[promptId]
    }
    process.stdout.write('.')
  }
  throw new Error('Timeout — ComfyUI tardó más de 5 min')
}

function extractImages(historyEntry) {
  const imgs = []
  for (const out of Object.values(historyEntry.outputs)) {
    if (out.images) imgs.push(...out.images)
  }
  return imgs
}

async function downloadImage(filename, type, subfolder, destPath) {
  const url = `${COMFYUI_URL}/view?filename=${encodeURIComponent(filename)}&type=${type}&subfolder=${encodeURIComponent(subfolder || '')}`
  const res  = await fetch(url)
  if (!res.ok) throw new Error(`Download error: ${res.status}`)
  const buf  = Buffer.from(await res.arrayBuffer())
  fs.writeFileSync(destPath, buf)
}

// ─── MAIN ─────────────────────────────────────────────────────────────────────
async function main() {
  console.log('\n╔══════════════════════════════════════════╗')
  console.log('║  ZAMI GIRLS — Prueba Fase 1: Rostros     ║')
  console.log('╚══════════════════════════════════════════╝\n')

  // Verificar ComfyUI
  console.log('▸ Verificando ComfyUI...')
  try {
    const res = await fetch(`${COMFYUI_URL}/system_stats`)
    if (!res.ok) throw new Error(res.status)
    const stats = await res.json()
    console.log(`  ✓ ComfyUI activo — VRAM: ${Math.round((stats.system?.vram_total||0)/1e9)}GB\n`)
  } catch (e) {
    console.error(`  ✗ No se puede conectar a ComfyUI: ${e.message}`)
    console.error(`  URL: ${COMFYUI_URL}`)
    process.exit(1)
  }

  // Crear carpeta de salida
  fs.mkdirSync(OUTPUT_DIR, { recursive: true })

  // Encolar 4 generaciones
  console.log('▸ Encolando 4 generaciones...')
  const jobs = []
  for (let i = 0; i < 4; i++) {
    const seed     = randomSeed()
    const workflow = buildWorkflow(PARAMS, seed)
    const promptId = await queuePrompt(workflow)
    jobs.push({ promptId, seed, index: i })
    await new Promise(r => setTimeout(r, 400))
  }

  // Esperar resultados
  console.log('\n▸ Esperando resultados (2 min aprox)...')
  const results = []
  for (const job of jobs) {
    try {
      const entry  = await pollResult(job.promptId, job.index)
      const images = extractImages(entry)
      // Tomar la última imagen (upscaled final)
      const img = images[images.length - 1]
      if (img) {
        const dest = path.join(OUTPUT_DIR, `rostro-${job.index + 1}-seed${job.seed}.png`)
        await downloadImage(img.filename, img.type, img.subfolder, dest)
        console.log(`  ✓ Guardada: ${path.relative(__dir, dest)}`)
        results.push({ index: job.index, seed: job.seed, file: dest, url: `${COMFYUI_URL}/view?filename=${encodeURIComponent(img.filename)}&type=${img.type}` })
      }
    } catch (err) {
      console.error(`  ✗ Job ${job.index + 1} falló: ${err.message}`)
    }
  }

  // Resumen
  console.log('\n╔══════════════════════════════════════════╗')
  console.log('║  RESULTADOS                              ║')
  console.log('╚══════════════════════════════════════════╝')
  for (const r of results) {
    console.log(`\n  Rostro ${r.index + 1} (seed: ${r.seed})`)
    console.log(`  Archivo: ${path.relative(__dir, r.file)}`)
    console.log(`  URL:     ${r.url}`)
  }
  console.log(`\n✓ ${results.length}/4 rostros generados en: ${OUTPUT_DIR}\n`)
}

main().catch(e => { console.error('\n✗ Error:', e.message); process.exit(1) })
